### PR TITLE
[foundation 3 of 3] DCOS-14745: Extract foundation

### PR DESCRIPTION
**Pull request by nLight**
ℹ️  License adds 200LOC, don't be intimidated by +333/-134 :)

## Local packages

- Introduced /packages dir for all the local packages we're not ready to completely extract.
As discussed offline with @orlandohohmeier it is considered beneficial since we don't need to mess around with `npm link` nor `npm install` as we change code inside those packages
- Rewired dcos-ui to catch everything from this dir and treat as an npm package while also including them in the `npm test` and `npm run lint` routines to make our life easier

## DCOSStore and ResourceTableUtil

- Moved from `foundation-ui` to DCOS since they are highly DCOS-UI specific

## PluginSDK

- Removed all the foundation bits from the PluginSDK, plugins now can just use the package
- Exposed DCOSStore and ResourceTableUtil

~Should go together with https://github.com/mesosphere/dcos-ui-plugins-private/pull/565~
~List of PRs: https://github.com/pulls?utf8=%E2%9C%93&q=is%3Aopen+is%3Apr+author%3AnLight+foundation~

🙇

### CODE REVIEW

**Comment by jfurrow** [comment]
When testing this, I'm receiving an error and a warning. Is this to be expected?

**Comment by nLight** [comment]
For some reason webpack wants to include LICENSE in the build. I suspect that's because foundation-ui package sits right there and maybe we need to exclude it somehow from the build.

Do we have webpack Gods on the team?

/cc @orlandohohmeier

File: foundation-ui/src/mount/MountService.js [hunk 1]
```diff
@@ -43,7 +43,7 @@ class MountService extends EventEmitter {
     let {components, instance, rank} = this;
 
     if (typeof component !== 'function') {
-      if (global.__DEV__) {
+      if (process.env.NODE_ENV !== 'production') {
```
Comment (mlunoe) [inline foundation-ui/src/mount/MountService.js]: ❤️

File: src/js/routes/index.js [hunk 2]
```diff
@@ -92,7 +92,7 @@ function getRoutes() {
 
   // Register packages
   indexRoute.children = indexRoute.children.concat(
-    RoutingService.getDefinition()
+    RoutingService.RoutingService.getDefinition()
```
Comment (mlunoe) [inline src/js/routes/index.js]: 🤔 Could we not get rid of the duplicated name?

File: src/js/utils/NavigationServiceUtil.js [hunk 3]
```diff
@@ -28,14 +29,14 @@ const NavigationServiceUtil = {
         const {path, category, childRoutes = []} = route;
         const primaryPath = `/${path}`;
 
-        NavigationService.registerCategory(category);
+        NavigationService.NavigationService.registerCategory(category);
```
Comment (mlunoe) [inline src/js/utils/NavigationServiceUtil.js]: Same here, could we not get rid of that duplicated name?

File: foundation-ui/index.js [hunk 4]
```diff
@@ -1,17 +1,13 @@
-import DCOSStore from './stores/DCOSStore';
-import ResourceTableUtil from './utils/ResourceTableUtil';
-import RoutingService from './routing';
-import NavigationService from './navigation';
-import {MountService} from './mount';
-import Mount from './mount/Mount';
+import RoutingService from './src/routing';
+import NavigationService from './src/navigation';
```
Comment (mlunoe) [inline foundation-ui/index.js]:
```js
import {NavigationService, EventTypes as NavigationEventTypes} from './src/navigation';
...
module.exports = {
  NavigationService,
  NavigationEventTypes,
};
```
Or something?

File: gulpfile.js [hunk 5]
```diff
@@ -25,7 +25,14 @@ gulp.task('ensureDevProxy', function () {
 
 function buildDependenciesArray(dependencies) {
   return Object.keys(dependencies).map(function (dependency) {
-    return dependency + '@' + dependencies[dependency];
+    var version = dependencies[dependency];
+
+    if (version.startsWith('./')) {
```
Comment (mlunoe) [inline gulpfile.js]: Let's get a comment here. Something like:
`// Module is local, include dependencies from its package.json`

File: foundation-ui/index.js (hunk 4 above)
Comment (nLight) [inline foundation-ui/index.js]: How about:

```
module.exports = {
  navigation: {...},
  routing: {...}
};
```

```
import {navigation: {EventTypes}} from 'foundation-ui';
```

Comment (nLight) [inline foundation-ui/index.js]: Technically those are sub-packages...

Comment (mlunoe) [inline foundation-ui/index.js]: Yup, let's do it!

**Comment by nLight** [comment]
@mlunoe @orlandohohmeier @jfurrow please give it another shot. I addressed all the comments and improved the original solution with some guidance from Orlando 🙇

**Comment by mlunoe** [comment]
This looks really good @nLight! I do get this error a lot navigating around:

**Comment by nLight** [comment]
@mlunoe good one! Looks like the change made the problem visible, we shouldn't ideally register a mount point in a constructor, rather using the react life cycle methods.

File: plugins/services/src/js/components/modals/NewCreateServiceModalServicePicker.js [hunk 6]
```diff
@@ -49,27 +49,30 @@ function JSONOption({columnClasses, onOptionSelect}) {
...
 
-    MountService.MountService.registerComponent(
-      MultiContainerOption,
-      'CreateService:ServicePicker:GridOptions',
-      2
-    );
+class NewCreateServiceModalServicePicker extends React.Component {
+  componentDidMount() {
+    OPTIONS.forEach(function (component, index) {
+      MountService.MountService.registerComponent(
```
Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalServicePicker.js]: 👍

**Comment by mlunoe** [comment]
Looking good! Works great!

**Comment by jfurrow** [comment]
It works great! Nice work!

**Pull request merged**